_drafts/kotlin-type-classes.md

```diff
@@ -0,0 +1,523 @@
+
+Last but not least, we can also use Kotlin [context receivers](https://blog.rockthejvm.com/kotlin-context-receivers/#2-dispatchers-and-receivers) to declare that a function needs a specific context to be executed. So, we can change the `process` function as follows:
+
+```kotlin
+context(ValidatorScope<T>)
+fun <T> process(toValidate: T) =
+    either {
+        val validated: T = toValidate.validate().bind()
+        // Do something with the validated object
+    }
+```
+
+To sum up, **context receivers are a way to add a context or a scope to a function without passing this context as an argument or without using the extension function mechanism**.
+
+What's next? Well, we still need to talk about the validation logic. We'll do it in the next section.
+
+## 5. Expanding the Validation Framework
+
+We introduced the `ValidatorScope<T>` interface in the previous section. We also saw how to use it to validate a DTO. However, we still need to talk about the validation logic. Let's do it now. We can use the type classes approach once again to solve the problem.
+
+First, we need to define the validation rules. We'll start with the `CreatePortfolioDTO` type. We want to validate the `userId` and the `amount` fields. The `userId` field must be a non-empty string, while the `amount` field must be a positive number. Let's define the validation rules as follows:
+
+```kotlin
+interface NonEmpty<T> {
+    fun T.nonEmpty(): Boolean
+}
+
+interface Positive<T : Number> {
+    fun T.positive(): Boolean
+}
+```
+
+As we can see, **the validation rules are generic**. This way, we can apply them to multiple types, exploiting ad-hoc polymorphisms again. And, yup, the above types are type classes. For the `CreatePortfolioDTO`, we need the following implementations:
+
+```kotlin
+val nonEmptyString = object : Required<String> {
+    override fun String.nonEmpty(): Boolean = this.isNotBlank()
+}
+
+val positiveDouble = object : Positive<Double> {
+    override fun Double.positive(): Boolean = this > 0.0
+}
+```
+
+Nothing will stop us from having multiple implementations of the same validation rule for the same type. For example, we can have the following implementation for the `List<String>` and `Int` types:
+
+```kotlin
+val nonEmptyList = object : NonEmpty<List<String>> {
+    override fun List<String>.nonEmpty(): Boolean = this.isNotEmpty()
+}
+
+val positiveInt = object : Positive<Int> {
+    override fun Int.positive(): Boolean = this > 0
+}
+```
+
+We need a type hierarchy implementing the errors that a single validation rule can generate. Let's define the following hierarchy:
+
+```kotlin
+sealed interface InvalidFieldError {
+
+    val field: String
+
+    data class MissingFieldError(override val field: String) : InvalidFieldError {
```

Comment:
I'd say we place them outside so that when we reference these types later we just use their shorthand names

Reply:
I want to leave it inside the interface. I saw it's a common way to declare sealed hierarchies since it adds context (the name o the interface).